---
title: "Phyloseq"
author: "Melia Hodson"
date: "11/17/2020"
output: github_document
---
# Intro

We explore the processed 2018 sequences using [phyloseq](https://joey711.github.io/phyloseq/)

# Install phyloseq

```{r}
# BiocManager::install("phyloseq")
```

```{r message=F, warning=F}
library(tidyverse)
library(phyloseq)
library(RColorBrewer)
library(readxl)
```

# Import Data

```{r}
count.tab <- read_rds("~/Desktop/Github/144L_demo/stuff/Input_Data/week6/seqtab-nochimtaxa.rds")
# Table of counts for each sequence in each sample
tax.tab <- read_rds("~/Desktop/Github/144L_demo/stuff/Input_Data/week6/taxa.rds")

metadata <- read_excel("~/Desktop/Github/144L_demo/stuff/Input_Data/week4/144L_2018_Exp_TOC.xlsx", sheet = "Metadata") %>% 
  select(Treatment, Bottle, Timepoint, DNA_SampleID)
sample.tab.noID <- read_rds("~/Desktop/Github/144L_demo/stuff/Output_Data/Week 4/2018_Experiment_Processed_DOC_BGE.Rmd")
sample.tab <- left_join(sample.tab.noID, metadata) %>% 
  drop_na(DNA_SampleID) %>% 
  column_to_rownames(var = "DNA_SampleID")

# Table that matches ASV to sequence
# sample.tab  %>% 
  # drop_na(DNA_SampleID) %>% 
  # column_to_rownames(var = "DNA_SampleID")

glimpse(sample.tab)

row.names(sample.tab)
```

# Phyloseq Object

We need to create a phyloseq object that merges all three datasets. Sometimes this doesn't work because of the format of the data files. Make sure all the sample names between the sampleinfo.txt and seqtab-nochimtaxa.txt are the same

```{r}
OTU = otu_table(count.tab, taxa_are_rows = T)
TAX = tax_table(tax.tab)
SAM = sample_data(sample.tab)
ps = phyloseq(OTU, TAX, SAM)
```

# Filter Sequences

We will filter out chloroplasts and mitochondria, because we only intended to amplify bacterial sequences. It's good to check you don't have anything lurking in the taxonomy table.

```{r}
sub_ps <- ps %>% 
  subset_taxa(Family != "mitochondria" & Order != "Chloroplast")
```

# Sample Summary

As a first analysis, we will look at the distribution of read counts from our samples

```{r echo=F,warning=F, message=F, fig.height=4, fig.width=6, fig.align="center"}
sample_sum_df <- data.frame(sum = sample_sums(sub_ps))

ggplot(sample_sum_df, aes(x = sum)) +
  geom_histogram(color = "black", fill = "royalblue3", binwidth = 4000) +
  ggtitle("Distribution of Sample Sequencing Depth") +
  xlab("Read Counts") +
  theme(axis.title.y = element_blank()) +
  theme_bw()
```
The Distribution shows the majority of the read counts lie between 2000 and 4000 read counts. However there is also a spike of reads closer to 6000 reads.

```{r}
# Mean, max, and min of sample read counts
summary(sample_sum_df)
```

# Beta Diversity

Beta diversity involves calculating metrics such as distances or dissimilarities based on pairwise comparisons of samples - they don't exist for a single sample, but rather only as metrics that relate samples to each other. i.e. beta diversity = patterns in community structure between samples

Since differences in sampling depths between samples can influence distances/dissimilarity metrics, we first need to somehow normalize the read depth across our samples.

## Subsample

We will rarefy (random subsample with replacement) the read depth of the samples first (scale to the smallest library size).

Read depth is an artifact of a machine made by a company in San Diego, not anything about your samples or biology. It is totally artifactual, and controlling for artifacts is critical in science. Subsampling randomly is the simplest way to control for this, and the question is whether this is the "best" way of controlling for it.

A strong reason to subsample is to standardize effort. The bottom line is that in all experimental design you should not be comparing things to which you devote different effort in resolution. You standardize effort.

The bigger your differential in mean(or median) read depth(reads/sample) between pre- and post-subsampling, the greater the "effect" on beta diversity.

We will subsample to the minimum read depth of all samples and not subsample. We'll then compare the mean reads pre- and post-subsampling and also compare beta diversity patterns.

```{r}
ps_min <- rarefy_even_depth(sub_ps, sample.size = min(sample_sums(sub_ps)))

mean(sample_sums(sub_ps)) # 31642 = median
mean(sample_sums(ps_min)) # 2432 = min
```
There is a big differential between the median (31642) and the minimum read depth (2432), which means the random subsampling could have a larger effect on our beta diversity. 

## NMDS

One of the best exploratory analyses for amplicon data is unconstrained ordinations. Here we will look at non-metric multidimensional scaling (NMDS) ordination of our full community samples. For NMDS plots it's important to set a seed since the starting positions of samplings in the algorithm is random.

We are running 2 sets of ordinations here, the first object (nmds) will be on the non-sampled dataset, and the second (nmds_min) will be on the sub-sampled dataset.

```{r}
set.seed(1)
# Ordinate
nmds <- ordinate(sub_ps, method = "NMDS", distance = "bray") # Stress = 0.07 - 0.15
```
```{r}
set.seed(1)
# Ordinate
nmds_min <- ordinate(ps_min, method = "NMDS", distance = "bray") # Stress = 0.07
```

We can now plot our ordinances now that they are stored in objects.

```{r echo=F, warning=F, message=F, fig.height=4, fig.width=6, fig.align="center"}
levels = c("Control", "Ash Leachate", "Mud Leachate", "Glucose_Nitrate_Phosphate") # "Bottle", "Vial"

nmds.plot <- plot_ordination(sub_ps, nmds, title = "NMDS for Non-samples Dataset") +
  geom_point(aes(fill = days, shape = factor(Treatment, levels = levels)), alpha = 0.6, stroke = 2, size = 4) +
  scale_shape_manual(values = c(21,22,23,24)) +
  scale_fill_gradient(low = "tan1", high = "royalblue3") +
  theme_bw()
# Removing one of the plotting layers (there are points within points)
nmds.plot$layers <- nmds.plot$layers[-1]
nmds.plot +
  facet_grid(~Location) +
  guides(fill = guide_colorbar(title = "Days"), shape = guide_legend(title = "Treatment"))
```
Based on this plot, for all the treatments their similarities were scattered at day 0. Approaching day 2 Ash Leachate and Mud Leachate both show a higher similarity to each other, with the control also nearby. The Glucose Nitrate Phosphate is much more dissimilar to the rest of the treatments at day 2. On day 4 all of the treatments show much more dissimilarity between the different communities with Glucose Nitrate Phosphate being the most dissimilar. 

```{r echo=F, warning=F, message=F, fig.height=4, fig.width=6, fig.align="center"}
nmds_min.plot <- plot_ordination(ps_min, nmds_min, title = "NMDS for Sub-sampled Dataset") +
  geom_point(aes(fill = days, shape = factor(Treatment, levels = levels)), alpha = 0.6, stroke = 2, size = 4) +
  scale_shape_manual(values = c(21,22,23,24)) +
  scale_fill_gradient(low = "tan1", high = "royalblue3") +
  theme_bw()
# Removing one of the plotting layers (there are points within points)
nmds_min.plot$layers <- nmds_min.plot$layers[-1]
nmds_min.plot +
  facet_grid(~Location) +
  guides(fill = guide_colorbar(title = "Days"), shape = guide_legend(title = "Treatment"))
```
Similar to the previous plot, the communities are more scattered at day 0 with the Mud Leachate treatment being the most similar. Day 2 shows the Mud Leachate and Control being the most similar and Glucose Nitrate Phosphate being most dissimilar. On Day 4 all the treatments are much more dissimilar compared to the previous days with Glucose Nitrate Phosphate being the most dissimilar. 

NMDS plots attempt to show ordinal distances between samples as accurately as possible in two dimensions. It is important to report the stress of these plots, because a high stress value means that the algorithm had a hard time representing the distances between samples in 2 dimensions. The stress of this plot was good - it was 0.04 (generally anything below 0.2 is considered acceptable)

Sub-sampling doesn't appear the pattern we see in beta diversity, so moving forward, we will focus on the sub-sampled dataset. 

# Alpha Diversity

We are going to calculate the Chao1 index for richness and the Shannon diversity index.

** It is important to note that the alpha diversity values are not interpreted as "real" numbers of anything (due to the nature of amplicon data), but they can still be useful as relative metrics of comparison. If Chao1 richness goes up, but Shannon diversity goes down, it indicates that the sample may have more ASVs but is dominated by a few of them **

We will use the subsampled library, which retains estimates of the species abundance of the real population while standardizing sampling effort.

```{r}
richness <- estimate_richness(ps_min, measures = c("Chao1", "Shannon")) %>% 
  rownames_to_column(., var = "DNA_ID") %>% 
  mutate_at(vars(DNA_ID), str_replace_all, "X144_","144_")
```

Let's add the sample metadata into this dataframe now.

```{r}
alphadiv <- left_join(richness, sample.tab %>% 
                        rownames_to_column(., var = "DNA_ID"))
```

```{r echo=F, warning=F, message=F, fig.height=6, fig.width=12, fig.align="center"}
# install.packages("ggpubr")
library(ggpubr)

pivot.data <- alphadiv %>% 
  select(Treatment, Bottle, Timepoint, days, Chao1, Shannon) %>% 
  pivot_longer(., cols = c(Chao1, Shannon), names_to = "measure", values_to = "est") %>% 
  left_join(., alphadiv %>% 
              select(Treatment, Bottle, Timepoint, days, se.chao1)) %>% 
  mutate(se.chao1 = ifelse(measure == "Chao1", se.chao1, NA))

alpha.plot <- ggboxplot(pivot.data, x = "Timepoint", y = "est", 
                        palette = c("royalblue3", "tan1"),
                        xlab = expression(italic(paste(""))),
                        ylab = expression(italic(paste("Alpha Diversity Measure"))),
                        add = "dotplot",
                        width = 0.7,
                        ggtheme = theme_bw()) +
  stat_compare_means(label.x = "4") +
  facet_grid(measure~ factor(Treatment, levels = levels),
             scales = "free")
alpha.plot

```

Boxes represent the 1.5 interquartile range, with the internal solid line are representing the median. Circles represent data points. p-values are reported the non-parametric two sample Wilcoxon test, which tests whether the means between two groups are equal (ns: p > 0.05, * :ps =< 0.05, ** :p =< 0.01).

Difference in the alpha diversity indexes among conditions were tested using pairwise Wilcoxon tests; p < 0.05 was considered the threshold significance for a difference between conditions.

From this plot we can see within the treatments that both the richness and diversity decrease by timepoint 4 which may correspond with stationary growth. The plots show that both the richness (via Chao1 index) of our samples and the overall diversity (via Shannon index) significantly changes. Notably, the n value is not ideal so alpha diversity measures likely won't hold as much weight. The Mud Leachate has the smallest change for both richness and overall diversity across timepoints. The Ash Leachate and the Glucose Nitrate Phosphate show similar results in both their overall diversity and richness, starting higher, dropping around timepoint 4, then increasing again. The Control stayed relatively constant across timepoints with much less variation compared to the treatments. 

```{r}
alpha.plot2 <- ggboxplot(pivot.data, x = "Timepoint", y = "est", 
                        palette = c("royalblue3", "tan1"),
                        xlab = expression(italic(paste(""))),
                        ylab = expression(italic(paste("Alpha Diversity Measure"))),
                        add = "dotplot",
                        width = 0.7,
                        ggtheme = theme_bw()) +
  stat_compare_means(label.x = "4") +
  facet_grid(rows = "measure", scales = "free")
alpha.plot2
```

# Who?

Which taxa were important? Which taxa were contributing to the change in community composition?

** Note: Recovered 16 rRNA gene copy numbers do not equal organism abundance. **

That said, we can generate a heat map of our samples showing us how the relative abundance of different taxonomic group change... potentially giving us a visual of which taxa are most important to the alpha and beta diversity patterns we observed.

First we're going to generate a custom table that will be easier to work with than a phyloseq object.

## Generate Relative Abundance

Our data currently shows number gene copies recovered, so we'll convert to percentages (relative abundance)

```{r warning=F}
ps_std <- transform_sample_counts(ps_min, function(x) x/sum(x))
# Extract the relative abundance table and coerce into dataframe
ps_std.tab <- as(otu_table(ps_std), "matrix")
ps_std.df = as.data.frame(ps_std.tab)
```

## Make Table

```{r}
custom.tab <- as.data.frame(tax.tab)
# Then combine the data frame
tax.df <- as.data.frame(tax.tab)

custom.tab <- tax.df %>% 
  rownames_to_column(., var = "asv") %>% 
  left_join(., ps_std.df %>%  rownames_to_column(., var = "asv")) %>% 
  mutate(pco = paste(Phylum, "_", Class, "_", Order)) %>% 
  select(-c(asv:Genus)) %>% 
  select(pco, everything()) %>% 
  group_by(pco) %>% 
  summarize_at(vars(contains("144")), sum, na.rm = T) %>% 
  ungroup()

# Save the row names and then make them into the column names
colnames <- custom.tab[, 1]

# Transpose the dataframe so we can merge with the sample info table
t_custom.tab <- as.data.frame(t(custom.tab[, -1]))

colnames(t_custom.tab) <- colnames$pco

# Merge
sweet.tab <- t_custom.tab %>% 
  rownames_to_column(., var = "sample") %>% 
  left_join(., sample.tab %>%  rownames_to_column(., var = "sample") %>% select(sample, Experiment, Bottle, Treatment, Timepoint, days, cells)) %>% 
  select(sample, Experiment:cells, everything())

relabund <- sweet.tab %>% 
  select(-c(sample:cells)) %>% 
  .[, colSums(.) > 0] %>% 
  .[, order(colSums(-.))] %>% 
  bind_cols(sweet.tab %>%  select(sample:cells), .)
```

## Heatmap

```{r echo=F, warning=F, message=F, fig.height=16, fig.width=10, fig.align="center"}
relaheat.data <- relabund %>% 
  select(-c(sample, Experiment, Bottle, days, cells)) %>% 
  pivot_longer(., -c(Treatment:Timepoint), names_to = "taxa", values_to = "relabund") %>% 
  separate(taxa, into = c("p", "c", "o"), sep = " _ ")

# install.packages("viridis")
library(viridis)

relaheat <- relaheat.data %>% 
  ggplot(aes(x = Timepoint, y = 0)) +
  geom_tile(aes(fill = relabund), color = "white") +
  scale_fill_viridis(option = "D") +
  labs(x = "Timepoint", y = "Order", fill = "Relative Abundance") +
  facet_grid(~factor(Treatment, levels = levels)) +
  theme_bw() +
  theme(axis.text.y = element_text(size =12),
        legend.position = "top") +
  guides(fill = guide_colorbar(barheight = 2, barwidth = 20, frame.colour = "black", ticks.colour = "black", ticks.linewidth = 1), color = F)

relaheat
```
It looks like the ASVs belonging to 

# Save and Knit

```{r}
saveRDS(sweet.tab, "~/Desktop/Github/144L_demo/stuff/Output_Data/Week 6/Custom_ASV_table.rds")
saveRDS(sub_ps, "~/Desktop/Github/144L_demo/stuff/Output_Data/Week 6/phyloseq_obj.rds")
saveRDS(ps_min, "~/Desktop/Github/144L_demo/stuff/Output_Data/Week 6/subsampled_phyloseq_obj.rds")
saveRDS(alphadiv, "~/Desktop/Github/144L_demo/stuff/Output_Data/Week 6/alphadiv.rds")
```

